core/src/org/samtools/htsjdk/core/io/ByteSource.java

```diff
@@ -0,0 +1,107 @@
+package org.samtools.htsjdk.core.io;
+
+import java.nio.ByteBuffer;
+
+/**
+ * Interface for classes that produce binary data (i.e. bytes) that can be decoded
+ * into other datatypes.
+ *
+ * Methods should throw exceptions if:
+ * - The underlying data source/stream/channel is closed
+ * - There are insufficient bytes available (after blocking) to read a single item of the datatype requested
+ *
+ * NOTES:
+ *   - Almost all the methods here could have default implementations that read via one of the readBytes() methods.
+ *   - Should think through how to deal with unsigned types
+ *   - Need to think how this would work with BlockCompressed data
+ */
+public interface ByteSource extends AutoCloseable {
+    /** Returns a description of the underlying source of the bytes, e.g. a path, stdin, a URL etc. */
+    String getDescription();
+
+    // Methods related to buffering/repositioning/skipping etc.
+    /** Returns true if the source has a known finite length, false otherwise. */
+    boolean hasLength();
+
+    /** Returns the length if {@link #hasLength()} is true, otherwise throws an exception. */
+    default long length() { throw new UnsupportedOperationException(); }
+
+    /** Returns the offset of the next byte that will be read by any operation. */
+    long position();
+
+    /** Returns true if the source can skip bytes without reading them. */
+    boolean canNavigateForward();
```

Comment:
I think `canSkip()` and `skip(long)` might be more familiar.  The meaning of `canNavigateForward` is sort of vague unless you read the comment.

Reply:
I'm happy to name them whatever.  Perhaps it could even be a single `canX()` method since I'm not sure when we'd have the ability to skip forwards (without reading) but not skip backwards or vice versa.